### PR TITLE
fix networkfs operation race

### PR DESCRIPTION
**Pull request by Vicente-Cheng**
#### Problem:
networkfilesystem operation race with the edge case (enable/disable very fast)

#### Solution:
Enhance the error handling of enable operation

#### Related Issue(s):
https://github.com/harvester/harvester/issues/7796

#### Test plan:
Please check the reproduce steps of https://github.com/harvester/harvester/issues/7796

### CODE REVIEW

**Review by tserong — DISMISSED** [review]
It took me a while to get my head around this one :-) Let's see if I understand correctly.

If the cause of the race is a very quick enable/disable, then we end up with:

- networkfs goes to "Disabling"
- sharemanager goes to "stopping"
- networkfs is then re-set to "Enabling" (before sharemanager is stopped), so the `if !isEnabling(networkFS) && lhShareMgr.Status.State == longhornv2.ShareManagerStateStopped` on line 116 fails, which menas the lhVA doesn't get set up (this is what causes it to get stuck)
- This new change here will pick up the case where sharemanager is either still stopping or stopped, and (because it sets the networkfs back to Disabled), will bounce us back to the start.

I haven't built and tested this myself, but the logic LGTM

File: pkg/controller/networkfilesystem/controller.go [hunk 1]
```diff
@@ -134,6 +134,22 @@ func (c *Controller) enableNetworkFS(networkFS *networkfsv1.NetworkFilesystem) (
...
+		}
+		if len(lhva.Spec.AttachmentTickets) == 0 {
+			logrus.Infof("The LHVA %s has no attachment tickets, reset the network filesystem status for retry", networkFS.Name)
+			networkFSCpy := networkFS.DeepCopy()
+			networkFSCpy.Status.State = networkfsv1.NetworkFSStateDisabled
+			networkFSCpy.Status.Status = networkfsv1.EndpointStatusNotReady
+			if !reflect.DeepEqual(networkFS, networkFSCpy) {
+				return c.NetworkFilsystems.UpdateStatus(networkFSCpy)
+			}
+		}
```
Comment (WebberHuang1118) [inline pkg/controller/networkfilesystem/controller.go]: Can you explain why this causes the network filesystem to retry? Also, would it be possible to factor out this part for better readability? Thanks!

**Comment by Vicente-Cheng** [comment]
> It took me a while to get my head around this one :-) Let's see if I understand correctly.
> 
> If the cause of the race is a very quick enable/disable, then we end up with:
> 
> * networkfs goes to "Disabling"
> * sharemanager goes to "stopping"
> * networkfs is then re-set to "Enabling" (before sharemanager is stopped), so the `if !isEnabling(networkFS) && lhShareMgr.Status.State == longhornv2.ShareManagerStateStopped` on line 116 fails, which menas the lhVA doesn't get set up (this is what causes it to get stuck)
> * This new change here will pick up the case where sharemanager is either still stopping or stopped, and (because it sets the networkfs back to Disabled), will bounce us back to the start.
> 
> I haven't built and tested this myself, but the logic LGTM

Thanks @tserong.

And yes, this is what we faced.
So, resetting the corresponding networks would help re-enable it.

File: pkg/controller/networkfilesystem/controller.go (hunk 1 above)
Comment (Vicente-Cheng) [reply pkg/controller/networkfilesystem/controller.go]: This is used to reset the networkfilesystem state.
So the next round we will enable (set LHVA) again, to ensure the sharemanager will bring up.

I could add some comments here to ensure the readability.

UPDATE: code updated, please recheck

**Review by tserong — DISMISSED** [review]

File: pkg/controller/networkfilesystem/controller.go [hunk 2]
```diff
@@ -134,6 +134,24 @@ func (c *Controller) enableNetworkFS(networkFS *networkfsv1.NetworkFilesystem) (
...
+		if len(lhva.Spec.AttachmentTickets) == 0 {
+			// this will reset the network filesystem status to disabled
+			// after reconcile, we will retry to enable the network filesystem
+			logrus.Infof("The LHVA %s has no attachment tickets, reset the network filesystem status for retry", networkFS.Name)
+			networkFSCpy := networkFS.DeepCopy()
+			networkFSCpy.Status.State = networkfsv1.NetworkFSStateDisabled
+			networkFSCpy.Status.Status = networkfsv1.EndpointStatusNotReady
+			if !reflect.DeepEqual(networkFS, networkFSCpy) {
+				return c.NetworkFilsystems.UpdateStatus(networkFSCpy)
+			}
```
Comment (WebberHuang1118) [inline pkg/controller/networkfilesystem/controller.go]: Have we considered factoring out this part for better readability?

**Review by WebberHuang1118 — DISMISSED** [review]
LGTM, just a nit, thanks.

File: pkg/controller/networkfilesystem/controller.go (hunk 2 above)
Comment (Vicente-Cheng) [reply pkg/controller/networkfilesystem/controller.go]: updated!

**Review by tserong — DISMISSED** [review]
LGTM.  The only other thought I had is that `setNetworkFSToDefault()` and `setNetworkFSToEnabling()` could be collapsed further into a single function which takes `state` as a parameter, something like this (untested):

```go
func setNetworkFSState(networkFS *networkfsv1.NetworkFilesystem, state networkfsv1.NetworkFSState) *networkfsv1.NetworkFilesystem {
        networkFSCpy := networkFS.DeepCopy()
        networkFSCpy.Status.Status = networkfsv1.EndpointStatusNotReady
        networkFSCpy.Status.State = state
        if state == networkfsv1.NetworkFSStateEnabling {
                networkFSCpy.Status.Type = targetNetFS
        }
        return networkFSCpy
}
```

...but I don't have a strong preference either way. The two separate functions as they are, are quite clear.

**Comment by Vicente-Cheng** [comment]
> LGTM. The only other thought I had is that `setNetworkFSToDefault()` and `setNetworkFSToEnabling()` could be collapsed further into a single function which takes `state` as a parameter, something like this (untested):
> 
> ```go
> func setNetworkFSState(networkFS *networkfsv1.NetworkFilesystem, state networkfsv1.NetworkFSState) *networkfsv1.NetworkFilesystem {
>         networkFSCpy := networkFS.DeepCopy()
>         networkFSCpy.Status.Status = networkfsv1.EndpointStatusNotReady
>         networkFSCpy.Status.State = state
>         if state == networkfsv1.NetworkFSStateEnabling {
>                 networkFSCpy.Status.Type = targetNetFS
>         }
>         return networkFSCpy
> }
> ```
> 
> ...but I don't have a strong preference either way. The two separate functions as they are, are quite clear.

I separate the setup function, and we could reuse it.
Thanks for the suggestion, please recheck it.

**Review by tserong — APPROVED** [review]
LGTM!

**Review by WebberHuang1118 — APPROVED** [review]
LGTM, thanks.

**Comment by Vicente-Cheng** [comment]
@mergifyio backport v1.6

**Comment by wheatdog** [comment]
@Mergifyio backport v1.5
